### PR TITLE
Add fuse driver to oi-userland

### PR DESCRIPTION
**Pull request by xen0l**
illumos FUSE driver from https://github.com/jurikm/illumos-fuse available in SFE, not part of oi-userland. I packaged the thing, but didn't do proper testing yet. Please review.

### CODE REVIEW

**Comment by xen0l** [comment]
The driver does not work when compiled by illumos gcc. This will need to investigated. Driver can be loaded when compiled with studio. As studio is deprecated, we should fix this and hopefully we might have a working fuse on illumos.

File: components/illumos/fuse/patches/compiler.patch [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+ CTFCONVERT = ctfconvert
+ LINT = lint
+diff -rup illumos-fusefs-Version-1.3.orig/kernel/amd64/Makefile illumos-fusefs-Version-1.3/kernel/amd64/Makefile
+--- illumos-fusefs-Version-1.3.orig/kernel/amd64/Makefile	2016-01-29 08:27:03.000000000 +0100
++++ illumos-fusefs-Version-1.3/kernel/amd64/Makefile	2016-04-15 01:17:49.005364646 +0200
+@@ -23,7 +23,7 @@
+ # Copyright 2008 Sun Microsystems, Inc.  All rights reserved.
+ # Use is subject to license terms.
+ 
+-CFLAGS=-xmodel=kernel -m64
```
Comment (xen0l) [inline components/illumos/fuse/patches/compiler.patch:20]: This line has to contain -ffreestanding -mcmodel=kernel -mno-red-zone. With that, the module can be compiled by illumos gcc and module can be loaded.

File: components/illumos/fuse/patches/compiler.patch [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
++CFLAGS= -m64
```
Comment (pyhalov) [inline components/illumos/fuse/patches/compiler.patch]: Is it 64-bit-only?

**Comment by pyhalov** [comment]
We'll also need libfuse in the same (or related) PR? At least we should somehow check that it's working, not just loading.

File: components/illumos/fuse/fuse.p5m [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+set name=info.classification value="org.opensolaris.category.2008:Drivers/Storage"
+set name=info.upstream-url value=https://github.com/jurim/illumos-fusefs
+set name=info.source-url value=$(COMPONENT_ARCHIVE_URL)
+set name=org.opensolaris.consolidation value=$(CONSOLIDATION)
+
+license CDDL.license license='CDDL'
+
+set name=variant.opensolaris.zone value="global"
+
+dir  path=usr
```
Comment (pyhalov) [inline components/illumos/fuse/fuse.p5m]: Remove dir actions

File: components/illumos/fuse/Makefile [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+COMPONENT_SRC=		$(COMPONENT_NAME)-Version-$(COMPONENT_VERSION)
+COMPONENT_ARCHIVE=	$(COMPONENT_SRC).zip
+COMPONENT_ARCHIVE_HASH=	sha256:29fd2dff679b4da58dbc0830730a6ac54d4a0fdeb91459d8341594db94f4c058
+COMPONENT_ARCHIVE_URL=	https://github.com/jurikm/illumos-fusefs/archive/Version-$(COMPONENT_VERSION).zip
+
+include $(WS_TOP)/make-rules/prep.mk
+include $(WS_TOP)/make-rules/justmake.mk
+include $(WS_TOP)/make-rules/ips.mk
+
+PATH=/opt/onbld/bin/i386/:/opt/onbld/bin:/usr/gnu/bin:/usr/bin:/usr/sbin:/sbin
```
Comment (pyhalov) [inline components/illumos/fuse/Makefile]: you could do it consistent :)

File: components/illumos/fuse/fuse.p5m [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+# http://www.illumos.org/license/CDDL.
+#
+
+#
+# Copyright 2016 Adam Stevko 
+#
+
+<transform file dir path=.*kernel.* -> default group sys>
+<transform file path=.*kernel/drv/.+\.conf -> default mode 0644>
+#<transform file path=.*kernel/drv/.+ -> default mode 0755>
```
Comment (pyhalov) [inline components/illumos/fuse/fuse.p5m]: Remove unnecessary comments

File: components/network/sshfs/Makefile [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+
+#
+# Copyright 2016 Adam Stevko. All rights reserved.
+#
+
+include ../../../make-rules/shared-macros.mk
+
+COMPONENT_NAME=		sshfs
+COMPONENT_VERSION=	2.7
+COMPONENT_FMRI=		network/$(COMPONENT_NAME)
```
Comment (xen0l) [inline components/network/sshfs/Makefile]: @pyhalov do you think that filesystem/fuse/sshfs is better and network/sshfs?

Comment (pyhalov) [inline components/network/sshfs/Makefile]: Do we have any filesystem/\* package ?

File: components/library/libfuse/Makefile [hunk 7]
```diff
@@ -0,0 +1,73 @@
...
+include $(WS_TOP)/make-rules/ips.mk
+
+PATH=/opt/onbld/bin/i386:/opt/onbld/bin:/usr/gnu/bin:/usr/bin:/bin:/usr/sbin:/sbin
+
+PATCH_LEVEL=1
+
+CFLAGS+="-fPIC"
+CFLAGS+="-DPIC"
+CFLAGS+="-std=c99"
+CFLAGS+="-g"
```
Comment (pyhalov) [inline components/library/libfuse/Makefile]: why do we need debug information here?

File: components/library/libfuse/Makefile [hunk 8]
```diff
@@ -0,0 +1,73 @@
...
+
+include $(WS_TOP)/make-rules/prep.mk
+include $(WS_TOP)/make-rules/justmake.mk
+include $(WS_TOP)/make-rules/ips.mk
+
+PATH=/opt/onbld/bin/i386:/opt/onbld/bin:/usr/gnu/bin:/usr/bin:/bin:/usr/sbin:/sbin
+
+PATCH_LEVEL=1
+
+CFLAGS+="-fPIC"
```
Comment (pyhalov) [inline components/library/libfuse/Makefile]: Perhaps, let's move common CFLAGS  in flags.patch?

File: components/network/sshfs/Makefile (hunk 6 above)
Comment (xen0l) [inline components/network/sshfs/Makefile]: We don't, leaving it as it

File: components/library/libfuse/Makefile (hunk 7 above)
Comment (xen0l) [inline components/library/libfuse/Makefile]: this was found in Makefile.com from source tarball

File: components/network/sshfs/Makefile [hunk 9]
```diff
@@ -0,0 +1,49 @@
...
+COMPONENT_PROJECT_URL=	https://github.com/libfuse/sshfs
+COMPONENT_ARCHIVE_URL=	https://github.com/libfuse/sshfs/releases/download/$(COMPONENT_SRC)/$(COMPONENT_ARCHIVE)
+COMPONENT_LICENSE =	GPLv2
+COMPONENT_LICENSE_FILE=	COPYING
+
+include $(WS_TOP)/make-rules/prep.mk
+include $(WS_TOP)/make-rules/configure.mk
+include $(WS_TOP)/make-rules/ips.mk
+
+#PKG_CONFIG_PATH.32=/usr/gnu/lib
```
Comment (pyhalov) [inline components/network/sshfs/Makefile]: remove these comments

**Pull request merged**